addons/base_vat/models/res_partner.py

```diff
@@ -140,50 +140,68 @@ def fix_eu_vat_number(self, country_id, vat):
+                partner_label = _("partner [%s]", partner.name)
+                msg = partner.build_vat_error_message(country and country.code.lower() or None, partner.vat, partner_label)
+                raise ValidationError(msg)
+
+    @api.model
+    def run_vat_test(self, vat_number, default_country):
+        """ Checks a VAT number, either syntactically or using VIES, depending
+        on the active company's configuration.
+        A first check is made by using the first two characters of the VAT as
+        the country code. It it fails, a second one is made using default_country instead.
+
+        :return: True if vat_number is correct
+        """
+        # Get company
         if self.env.context.get('company_id'):
             company = self.env['res.company'].browse(self.env.context['company_id'])
         else:
             company = self.env.company
+
+        # Get check function: either simple syntactic check or call to VIES service
         eu_countries = self.env.ref('base.europe').country_ids
-        for partner in self:
-            if not partner.vat:
-                continue
-            #check with country code as prefix of the TIN
-            vat_country, vat_number = self._split_vat(partner.vat)
-            if company.vat_check_vies and partner.commercial_partner_id.country_id in eu_countries:
-                # force full VIES online check
-                check_func = self.vies_vat_check
-            else:
-                # quick and partial off-line checksum validation
-                check_func = self.simple_vat_check
-            if not check_func(vat_country, vat_number):
-                #if fails, check with country code from country
-                country_code = partner.commercial_partner_id.country_id.code
-                if country_code:
-                    if not check_func(country_code.lower(), partner.vat):
-                        msg = partner._construct_constraint_msg(country_code.lower())
-                        raise ValidationError(msg)
-
-    def _construct_constraint_msg(self, country_code):
-        self.ensure_one()
-        vat_no = "'CC##' (CC=Country Code, ##=VAT Number)"
-        vat_no = _ref_vat.get(country_code) or vat_no
+        if company.vat_check_vies and default_country in eu_countries:
+            check_func = self.vies_vat_check
+        else:
+            check_func = self.simple_vat_check
+
+        # First check with country code as prefix of the TIN
+        vat_country_code, vat_number_split = self._split_vat(vat_number)
+        if check_func(vat_country_code, vat_number_split):
+            return vat_country_code
+
+        # If it fails, check with default_country (if it exists)
+        if default_country and check_func(default_country.code.lower(), vat_number):
+            return default_country.code.lower()
+
+        return False
+
+    @api.model
+    def build_vat_error_message(self, country_code, wrong_vat, record_label):
```

Comment:
It's called from the account.fiscal.position constraint checking the foreign VAT. So, api.model makes sense here, since we don't have a record to call it on in that context.

Reply:
Indeed, but my point is that it does nothing more than a comment or docstring. Not asking to remove it, but I wouldn't have put it since we don't have guidelines for that and part of the framework team seems to want to remove it when it is useless.